just_audio/ios/Classes/SwiftPlayer.swift

```diff
@@ -0,0 +1,499 @@
+        player.addAudioSource(audioSequence)
+
+        let effects = audioSequence.sequence.map { audioSource in
+            audioSource.effects
+        }.flatMap { $0 }
+
+        audioSourcesAudioEffects = effects.reduce(into: audioSourcesAudioEffects) { partialResult, audioEffect in
+            partialResult[UUID().uuidString] = audioEffect
+        }
+
+        try onSetShuffleOrder(request: request)
+    }
+
+    func onSetShuffleOrder(request: [String: Any?]) throws {
+        guard let shuffleOrder = request["shuffleOrder"] as? [Int] else {
+            return
+        }
+
+        try player.shuffle(at: 0, inOrder: shuffleOrder)
+    }
+
+    func onAudioEffectSetEnabled(_ request: [String: Any]) throws {
+        let rawType = request["type"] as! String
+        let enabled = request["enabled"] as! Bool
+
+        if rawType == "DarwinEqualizer" {
+            if enabled {
+                try player.activateEqualizerPreset(at: 0)
+            } else {
+                try player.resetGains()
+            }
+
+            return
+        }
+
+        let type = DarwinAudioEffect(rawValue: rawType)!
+
+        let effect = try globalAudioEffects.values.first(where: { effect in
+            try effect.type == type
+        })
+
+        effect?.setBypass(enabled)
+    }
+
+    func onEqualizerBandSetGain(_ request: [String: Any]) throws {
+        let bandIndex = request["bandIndex"] as! Int
+        let gain = request["gain"] as! Double
+        try player.tweakEqualizerBandGain(band: bandIndex, gain: Float(gain))
+    }
+}
+
+// MARK: - SwiftPlayer streams
+
+@available(iOS 13.0, *)
+extension SwiftPlayer {
+    func subscribeToPlayerEvents() {
+        guard let safePlayer = player else {
+            return
+        }
+
+        // data channel
+        let outputPublishers = Publishers.CombineLatest(
+            safePlayer.$outputAbsolutePath,
+            safePlayer.$outputWriteError
+        )
+
+        let playerInfoPublishers = Publishers.CombineLatest3(
+            safePlayer.$isPlaying,
+            safePlayer.$volume,
+            safePlayer.$speed
+        )
+
+        let sideInfosPublishers = Publishers.CombineLatest(
+            safePlayer.$loopMode,
+            safePlayer.isShuffling
+        )
+
+        Publishers.CombineLatest3(
+            outputPublishers,
+            playerInfoPublishers,
+            sideInfosPublishers
+        )
+        .map { outputAbsoluteInfo, playerInfoPublishers, sideInfosPublishers in
+            DataChannelMessage(
+                outputAbsolutePath: outputAbsoluteInfo.0,
+                outputError: outputAbsoluteInfo.1,
+                playing: playerInfoPublishers.0,
+                volume: playerInfoPublishers.1,
+                speed: playerInfoPublishers.2,
+                loopMode: sideInfosPublishers.0,
+                shuffleMode: sideInfosPublishers.1
+            )
+        }
+        .removeDuplicates()
+        .receive(on: DispatchQueue.main)
```

Comment:
```suggestion 
```
since we are not touching any ui, there is no need to receive in the main thread, please apply where necessary

Reply:
Quando si emette un evento a flutter è necessario spostarsi sul Main Thread come in questo caso

Reply:
ok, let's keep it